A80-pid.md

```diff
@@ -0,0 +1,312 @@
+
+### LB Policy Config and Parameters
+
+The `pid` LB policy config will be as follows.
+
+```textproto
+message LoadBalancingConfig {
+  oneof policy {
+    PIDLbConfig pid = 20 [json_name = "pid"];
+  }
+}
+
+message PIDLbConfig {
+  // Configuration for the WRR load balancer as defined in [gRFC A58][A58].
+  // The PID balancer is an extension of WRR and all settings applicable to WRR also apply to PID identically.
+  WeightedRoundRobinLbConfig wrr_config = 1;
+
+  // Threshold beyond which the balancer starts considering the ErrorUtilizationPenalty.
+  // This helps avoid oscillations in cases where the server experiences a very high and spiky error rate.
+  // We avoid eliminating the error_utilization_penalty entirely to prevent redirecting all traffic to an instance 
+  // that has low CPU usage but rejects all requests. Default is 0.5.
+  google.protobuf.FloatValue error_utilization_threshold = 2;
+
+  // Controls the convergence speed of the PID controller. Higher values accelerate convergence but may induce oscillations,
+  // especially if server load changes more rapidly than the PID controller can adjust. Oscillations might also occur due to
+  // significant delays in load report propagation or extremely spiky server load. To mitigate spiky loads, server owners should
+  // employ a moving average to smooth the load reporting. Default is 0.1.
+  google.protobuf.FloatValue proportional_gain = 2;
+
+  // Adjusts the smoothness of the PID controller convergence. Higher values enhance smoothness but can decelerate convergence.
+  // Default is 1.
+  google.protobuf.FloatValue derivative_gain = 4;
+
+  // Maximum allowable weight. Weights proposed by the PID controller exceeding this value will be capped.
+  // This prevents infinite weight growth, which could occur if only a subset of clients uses PID and increasing weights
+  // no longer effectively corrects the imbalance. Default is 10.
+  google.protobuf.FloatValue max_weight = 5;
+
+  // Minimum allowable weight. Weights proposed by the PID controller falling below this value will be capped.
+  // This prevents weights from dropping to zero, which could occur if only a subset of clients uses PID and decreasing weights
+  // no longer effectively corrects the imbalance. Default is 0.1.
+  google.protobuf.FloatValue min_weight = 6;
+}
+```
+
+### PID controller
+
+A PID controller is a control loop feedback mechanism that continuously calculates an error value as the difference between a desired setpoint (`referenceSignal`) and a measured process variable (`actualSignal`). It then applies a correction based on proportional, integral, and derivative terms (denoted P, I, and D respectively), hence the name.
+
+In our implementation, we will not be using the integral part. The integral component is useful for speeding up convergence when the `referenceSignal` changes sharply. In our case, we will be converging the load on the subchannels to a mean value, which is mostly stable.
```

Comment:
Not extremely relevant, but I think this is traditionally called a PD controller when you omit the integral component.

I recall the integral component taking "history" into account. That is, it is proportionally weighted to how long it takes the controller to converge on a set point. I'm curious did you end up not using this because it caused too much oscillation?

Reply:
Yes, we could use integral part but we need to clamp the integral to not take into account too much of the history (or use windowed integral) This is similar to using moving average on the server side. Server side moving average is better IMO as it protects the server and gives server owners control over how much of the load history should be taken into account (which depends on the server properties, like how spiky is the server load)